msg-socket/src/req/socket.rs

```diff
@@ -42,6 +46,11 @@ impl<T: ClientTransport> ReqSocket<T> {
     }
 
     pub async fn request(&self, message: Bytes) -> Result<Bytes, ReqError> {
+        if self.active_requests.load(Ordering::SeqCst) >= self.options.max_active_requests {
+            return Err(ReqError::TooManyRequests);
+        } 
+        self.active_requests.fetch_add(1, Ordering::SeqCst);
```

Comment:
The actual incrementing should happen in the driver, since there's a lot that can go wrong between now and when the actual request is sent. The driver is much closer to the wire, and can make sure failed requests don't result in runaway counters.